### ISSUE TITLE
Class 'DataStore' incorrectly extends base class 'Container' - Typescript

### ISSUE THREAD

**Issue by miguelramos** [js-data/js-data#335]
When using on typescript (branch - v3.0.0-beta.6) i've got this error from typecript:

```
es6 class

export class StoreProvider extends DataStore {
  constructor(config?: any) {
    super(config);
  }
}

node_modules/js-data/dist/js-data.d.ts(219,14): error TS2415: Class 'DataStore' incorrectly extends base class 'Container'.

Types of property 'as' are incompatible.
    Type '(name: string) => Mapper | LinkedCollection' is not assignable to type '(name: string) => Mapper'.
      Type 'Mapper | LinkedCollection' is not assignable to type 'Mapper'.
        Type 'LinkedCollection' is not assignable to type 'Mapper'.
          Property '_adapters' is missing in type 'LinkedCollection'.
```
I'm using typescript@next version (1.9 at the moment) and node version v4.2.2.

**Comment by jmdobry** [comment]
I'm not a TypeScript expert, how should js-data.d.ts be changed to fix this?

**Comment by nickescallon** [comment]
@jmdobry when overriding a (public or protected) property, the override has to be assignable to the property it is overriding. 

in this case, the return value of `as` in `DataStore` (`Mapper|LinkedCollection`) is not assignable to `Mapper` (the return type that is being overriden).

in order to fix this, we could:
-  alter return type of `Container.as` to be `Mapper|LinkedCollection`, and skip defining the override.

- alter the return type of `Container.as` to `any`, and keep the more explicitly typed override

- if there are any differences in the call signatures, typescript would support the override, but I don't think that's the case here.

I'm not certain that this is the "best" way to fix this, but it does fix it. Happy to put up a PR if you like any of the options above.

**Comment by jmdobry** [comment]
I like option #2. PRs are welcome, especially on the `.d.ts` file.

**Comment by jmdobry** [comment]
Fixed by #347
